### PR TITLE
SignMessage rpc

### PR DESCRIPTION
**Pull request by bitcoin-coder-bob**
requested by: https://github.com/arkade-os/arkd/pull/904#pullrequestreview-3883920283

  - Proto: Added SignMessage RPC to SignerService with request/response messages                                                                                                                           
  - Generated files: Updated .pb.go, .pb.rgw.go, _grpc.pb.go, and OpenAPI spec
  - Port interface: Added SignMessage to SignerService interface
  - Infrastructure client: Implemented SignMessage on signerClient
  - Wallet types: Added SignMessage to WalletService interface
  - Wallet service: Implemented SignMessage using schnorr signing
  - gRPC handler: Added SignMessage handler to signerHandler
  - Tests: New test file with comprehensive SignMessage tests + fixture

## Summary by CodeRabbit

* **New Features**
  * Added message signing endpoint (POST /v1/sign-message) accessible via HTTP and gRPC; accepts hex-encoded messages and returns hex-encoded signatures. Input/output include validation (hex format, expected signature length) and clear errors for invalid payloads or missing signer key.

* **Tests**
  * Added fixture-driven tests covering valid signings, edge cases (empty, large, fixed-size), key selection, and missing-signer error.

### CODE REVIEW

File: internal/infrastructure/signer/client.go [hunk 1]
```diff
@@ -99,3 +99,15 @@ func (c *signerClient) SignTransactionTapscript(
 	}
 	return resp.GetSignedTx(), nil
 }
+
+func (c *signerClient) SignMessage(
+	ctx context.Context, message []byte,
```
Comment (IsaqueFranklin) [inline internal/infrastructure/signer/client.go:134]: I noticed that in `SignMessage` here returns `[]byte`, but in `SignTransactionTapscript` it returns a hex-encoded `string`. Is there a specific reason for this? Should we maintain consistency here and also return a hex-encoded 'string'? Just a thought on keeping the patterns consistent.

File: pkg/arkd-wallet/core/application/wallet/service.go [hunk 2]
```diff
@@ -727,6 +727,19 @@ func (w *wallet) LoadSignerKey(ctx context.Context, prvkey *btcec.PrivateKey) er
 	return nil
 }
 
+func (w *wallet) SignMessage(ctx context.Context, message []byte) ([]byte, error) {
+	if w.SignerKey == nil {
+		return nil, fmt.Errorf("signer key not loaded")
+	}
+
+	msgHash := chainhash.HashB(message)
```
Comment (IsaqueFranklin) [inline pkg/arkd-wallet/core/application/wallet/service.go:952]: This is nice and well implemented. 

Not sure if it applies here, however, should we consider using tagged hashes (BIP-340), meaning adding a context tag to the message before hashing? This prevents a message signature to be maliciously reused. Like, an attacker could create a message whose raw bytes are the same to a valid Bitcoin tx, the resulting signature would be valid on-chain. The context tag prevents that.

**Review by IsaqueFranklin — APPROVED** [review]
Overall it looks very solid!

I ran the unit tests provided and validated the changes on my machine, specifically `TestSignMessage` and `TestSignMessage_ConsistentWithSchnorrVerify` and they passed just fine.

Sorry for the nits 😅

cACK.

File: internal/infrastructure/signer/client.go (hunk 1 above)
Comment (bitcoin-coder-bob) [reply internal/infrastructure/signer/client.go:134]: `SignTransactionTapscript` returns a signed transaction (hex is the standard Bitcoin tx encoding), while `SignMessage` returns a raw Schnorr signature (64 bytes). The proto definition uses bytes for the signature field, so []byte is the natural Go type. Returning raw bytes is correct here and more flexible (callers can hex-encode if they want). I'd say this is fine as-is.

File: pkg/arkd-wallet/core/application/wallet/service.go (hunk 2 above)
Comment (bitcoin-coder-bob) [reply pkg/arkd-wallet/core/application/wallet/service.go:952]: `SignMessage` is a general-purpose, baking a specific tag into it would make it less flexible and would require all verifiers to know the tag. If domain separation is ever needed, the caller (like the indexer) can prepend its own tag to the message before calling `SignMessage`. I dont think in the context we use `SignMessage` that we need the tagged hashes.

Comment (IsaqueFranklin) [reply pkg/arkd-wallet/core/application/wallet/service.go:952]: Okay, I agree.

File: api-spec/protobuf/signer/v1/service.proto [hunk 3]
```diff
@@ -54,4 +60,11 @@ message SignTransactionTapscriptRequest {
...
 message SignTransactionTapscriptResponse {
   string signed_tx = 1;
+}
+
+message SignMessageRequest {
+  bytes message = 1;
+}
+message SignMessageResponse {
+  bytes signature = 1;
 }
```
Comment (louisinger) [inline api-spec/protobuf/signer/v1/service.proto:85]:
```suggestion
message SignMessageRequest {
  string message = 1;
}
message SignMessageResponse {
  string signature = 1;
}
```

let's use hex-encoded string instead of "bytes" to be consistent with other messages + make the encoding of REST API explicit (bytes is base64 I think)

File: pkg/arkd-wallet/core/application/wallet/service_test.go [hunk 4]
```diff
@@ -0,0 +1,219 @@
...
+
+	message := []byte("test message")
+	signature, err := w.SignMessage(ctx, message)
+
+	require.Error(t, err)
+	require.Nil(t, signature)
+	require.Contains(t, err.Error(), "signer key not loaded")
+}
+
+func TestSignMessage_DifferentKeysProduceDifferentSignatures(t *testing.T) {
```
Comment (louisinger) [inline pkg/arkd-wallet/core/application/wallet/service_test.go]: remove ? it seems not needed to me and can be covered by JSON test cases

File: pkg/arkd-wallet/core/application/wallet/service_test.go [hunk 5]
```diff
@@ -0,0 +1,219 @@
...
+	parsedSig2, err := schnorr.ParseSignature(sig2)
+	require.NoError(t, err)
+	require.True(t, parsedSig2.Verify(msgHash, privKey2.PubKey()))
+
+	// Cross-verification should fail
+	require.False(t, parsedSig1.Verify(msgHash, privKey2.PubKey()), "sig1 should not verify with key2")
+	require.False(t, parsedSig2.Verify(msgHash, privKey1.PubKey()), "sig2 should not verify with key1")
+}
+
+func TestSignMessage_DifferentMessagesProduceDifferentSignatures(t *testing.T) {
```
Comment (louisinger) [inline pkg/arkd-wallet/core/application/wallet/service_test.go]: same, remove, move to fixtures ?

File: pkg/arkd-wallet/core/application/wallet/service_test.go [hunk 6]
```diff
@@ -0,0 +1,219 @@
...
+	require.True(t, parsedSig1.Verify(chainhash.HashB(message1), pubKey))
+	require.False(t, parsedSig1.Verify(chainhash.HashB(message2), pubKey))
+
+	parsedSig2, err := schnorr.ParseSignature(sig2)
+	require.NoError(t, err)
+	require.True(t, parsedSig2.Verify(chainhash.HashB(message2), pubKey))
+	require.False(t, parsedSig2.Verify(chainhash.HashB(message1), pubKey))
+}
+
+func TestSignMessage_ConsistentWithSchnorrVerify(t *testing.T) {
```
Comment (louisinger) [inline pkg/arkd-wallet/core/application/wallet/service_test.go]: remove ? we do run Verify in TestSignMessage

File: pkg/arkd-wallet/core/application/wallet/fixtures/service_fixtures.json [hunk 7]
```diff
@@ -0,0 +1,48 @@
+{
```
Comment (louisinger) [inline pkg/arkd-wallet/core/application/wallet/fixtures/service_fixtures.json]: let's name the file `signmessage_fixtures.json` + rename "fixtures" to "testdata" (to be consistent with other fixture format)

also let's follow same format for JSON: 

```
{
    "valid": ....
     "invalid": .... 
}
```

(we need invalid test cases btw)

**Review by louisinger — CHANGES_REQUESTED** [review]

File: api-spec/protobuf/signer/v1/service.proto (hunk 3 above)
Comment (bitcoin-coder-bob) [reply api-spec/protobuf/signer/v1/service.proto:85]: @IsaqueFranklin asked the same: https://github.com/arkade-os/arkd/pull/954#discussion_r2921865945 to which I responded. Do you still think based off my response string is preferable?

File: internal/core/ports/signer.go [hunk 8]
```diff
@@ -13,4 +13,5 @@ type SignerService interface {
 	SignTransactionTapscript(
 		ctx context.Context, partialTx string, inputIndexes []int, // inputIndexes == nil means sign all inputs
 	) (string, error)
+	SignMessage(ctx context.Context, message []byte) ([]byte, error)
```
Comment (altafan) [inline internal/core/ports/signer.go:24]: For consitency

```suggestion
	SignMessage(ctx context.Context, message string) (string, error)
```

File: api-spec/protobuf/signer/v1/service.proto (hunk 3 above)
Comment (altafan) [reply api-spec/protobuf/signer/v1/service.proto:85]: @bitcoin-coder-bob let's move to `string`: 1. consistency compared to all others already existing rpcs 2. since we're mapping rpcs to rest with gateway, if we use bytes, they end up being encoded in hex or base64 depending on the gateway dep we choose, therefore to keep control on what happens, we decided to use string everywhere

File: pkg/arkd-wallet/core/application/wallet/fixtures/service_fixtures.json (hunk 7 above)
Comment (bitcoin-coder-bob) [reply pkg/arkd-wallet/core/application/wallet/fixtures/service_fixtures.json]: addressed in: [ebe18da](https://github.com/arkade-os/arkd/pull/954/commits/ebe18da1826c936e559d08cfe118c2e832d556a0)

File: pkg/arkd-wallet/core/application/wallet/service_test.go (hunk 6 above)
Comment (bitcoin-coder-bob) [reply pkg/arkd-wallet/core/application/wallet/service_test.go]: addressed in: [ebe18da](https://github.com/arkade-os/arkd/pull/954/commits/ebe18da1826c936e559d08cfe118c2e832d556a0)

File: pkg/arkd-wallet/core/application/wallet/service_test.go (hunk 5 above)
Comment (bitcoin-coder-bob) [reply pkg/arkd-wallet/core/application/wallet/service_test.go]: addressed in: [ebe18da](https://github.com/arkade-os/arkd/pull/954/commits/ebe18da1826c936e559d08cfe118c2e832d556a0)

File: pkg/arkd-wallet/core/application/wallet/service_test.go (hunk 4 above)
Comment (bitcoin-coder-bob) [reply pkg/arkd-wallet/core/application/wallet/service_test.go]: addressed in: [ebe18da](https://github.com/arkade-os/arkd/pull/954/commits/ebe18da1826c936e559d08cfe118c2e832d556a0)

File: internal/core/ports/signer.go (hunk 8 above)
Comment (bitcoin-coder-bob) [reply internal/core/ports/signer.go:24]: got a couple reviews saying similar: 
https://github.com/arkade-os/arkd/pull/954#discussion_r2921916063  and  https://github.com/arkade-os/arkd/pull/954#discussion_r2930811426  to which I replied: https://github.com/arkade-os/arkd/pull/954#discussion_r2927915842

what do we think, still go back to string or keep []byte?

Comment (bitcoin-coder-bob) [reply internal/core/ports/signer.go:24]: done in: [31e5b44](https://github.com/arkade-os/arkd/pull/954/commits/31e5b44c9ec7f025541e4fa0d686f27d8d684fa9)

File: api-spec/protobuf/signer/v1/service.proto (hunk 3 above)
Comment (bitcoin-coder-bob) [reply api-spec/protobuf/signer/v1/service.proto:85]: done in: [31e5b44](https://github.com/arkade-os/arkd/pull/954/commits/31e5b44c9ec7f025541e4fa0d686f27d8d684fa9)

File: api-spec/openapi/swagger/signer/v1/service.openapi.json [hunk 9]
```diff
@@ -184,6 +224,24 @@
...
+      },
+      "SignMessageResponse": {
+        "title": "SignMessageResponse",
+        "type": "object",
+        "properties": {
+          "signature": {
+            "type": "string"
+          }
+        }
+      },
```
[the remark this answers is not in the document]
Comment (bitcoin-coder-bob) [reply api-spec/openapi/swagger/signer/v1/service.openapi.json:270]: seems like a good suggestion, will add

[the remark this answers is not in the document]
Comment (bitcoin-coder-bob) [reply api-spec/openapi/swagger/signer/v1/service.openapi.json:270]: done in: [db3e733](https://github.com/arkade-os/arkd/pull/954/commits/db3e733789afa84020350b325717f1f4b4c5ca2f)

[the remark this answers is not in the document]
Comment (bitcoin-coder-bob) [reply api-spec/openapi/swagger/signer/v1/service.openapi.json:270]: done in: [fa47e25](https://github.com/arkade-os/arkd/pull/954/commits/fa47e2516f808f4ddef6e27c2a84ad29a16bf214)

File: pkg/arkd-wallet/core/application/wallet/testdata/signmessage_fixtures.json [hunk 10]
```diff
@@ -0,0 +1,54 @@
...
+      "name": "key_1",
+      "private_key_hex": "0000000000000000000000000000000000000000000000000000000000000001",
+      "public_key_hex": "0279be667ef9dcbbac55a06295ce870b07029bfcdb2dce28d959f2815b16f81798"
+    },
+    {
+      "name": "key_2",
+      "private_key_hex": "0000000000000000000000000000000000000000000000000000000000000002",
+      "public_key_hex": "02c6047f9441ed7d6d3045406e95c07cd85c778e4b8cef3ca7abac09b95c709ee5"
+    }
+  ],
```
Comment (louisinger) [inline pkg/arkd-wallet/core/application/wallet/testdata/signmessage_fixtures.json]: can we move the keys inside the test case itself ? to avoid having "getTestKey" logic and simplify the tests ?

**Review by ghost — COMMENTED** [review]
## 🔍 Arkana Review — SignMessage RPC

### Summary
Adds a new `SignMessage` RPC to the signer service that signs arbitrary messages with the operator's Schnorr key. Exposes via gRPC and REST (`POST /v1/sign-message`).

### Security

**⚠️ Access control on signer service endpoint**: The `POST /v1/sign-message` endpoint accepts any hex-encoded message and signs it with the operator's private key. This is fine if the signer service is only accessible internally (operator-to-signer communication), but could be dangerous if any part of the API surface is exposed to untrusted callers. Worth confirming the signer service is not reachable from client-facing ports.

**✅ Message hashing**: The implementation correctly hashes the message with `chainhash.HashB(message)` (double-SHA256) before passing to `schnorr.Sign`. This prevents signing raw transaction sighashes that an attacker might try to sneak through — the double-SHA256 wrapping ensures the signed digest can't be confused with a transaction signature.

**ℹ️ No message length limit**: The protobuf pattern `^(?:[0-9a-fA-F]{2})*$` validates hex format but doesn't limit length. For a signing endpoint, this is generally fine since the message is hashed anyway, but a max length would be a sensible defense-in-depth measure against abuse.

### Implementation
- Proto definition, generated code, Go handler, and wallet implementation all look correct and consistent.
- Test fixtures cover basic cases including empty message, various key lengths, and the no-key error path.
- Schnorr signature size validation in the response (`^[0-9a-fA-F]{128}$`, min/max 128 chars = 64 bytes) is correct.

### Nit
- `service.proto` — missing trailing newline at EOF.

Overall clean and well-structured. The main question is ensuring the signer service's network exposure is limited to trusted internal callers.

File: pkg/arkd-wallet/core/application/wallet/testdata/signmessage_fixtures.json (hunk 10 above)
Comment (bitcoin-coder-bob) [reply pkg/arkd-wallet/core/application/wallet/testdata/signmessage_fixtures.json]: done in: [343a8ae](https://github.com/arkade-os/arkd/pull/954/commits/343a8ae161a3e69eb7cc6aadf7a59071751f094f)

File: api-spec/protobuf/signer/v1/service.proto [hunk 11]
```diff
@@ -54,4 +60,19 @@ message SignTransactionTapscriptRequest {
...
+    pattern: "^(?:[0-9a-fA-F]{2})*$"
+  }];
+}
+message SignMessageResponse {
+  // hex-encoded Schnorr signature
+  string signature = 1 [(meshapi.gateway.openapi_field) = {
+    pattern: "^[0-9a-fA-F]{128}$"
+    min_length: 128
+    max_length: 128
+  }];
```
[the remark this answers is not in the document]
Comment (bitcoin-coder-bob) [reply api-spec/protobuf/signer/v1/service.proto:84]: done in: [d5c2963](https://github.com/arkade-os/arkd/pull/954/commits/d5c296364ce7128e33901981c0a9aa0e237c8d07)

**Review by ghost — COMMENTED** [review]
## 🔍 Arkana Review — arkd#954

**Summary:** Adds a `SignMessage` RPC to the `SignerService`. Accepts a hex-encoded message, signs it with Schnorr (via `chainhash.HashB` then `schnorr.Sign`), and returns the hex-encoded 64-byte signature.

### 🔒 Security — Key observations

1. **Signing scheme:** The message is hashed with `chainhash.HashB` (double SHA-256) before being passed to `schnorr.Sign`. This means the RPC signs `SHA256(SHA256(message))`, which is Bitcoin-standard. **Good** — no risk of signing raw transaction hashes that could be replayed as actual transactions, since the hash domain is distinct.

2. **No authentication boundary visible in this diff** — the `SignerService` is wired through the macaroon auth interceptor (visible in the interceptor chain in `interceptor.go`), so this RPC inherits the existing auth requirements. Anyone with a valid macaroon can call `SignMessage`. This is appropriate since they can already call `SignTransaction`.

3. **No message length limit on the input** — the proto schema only validates hex format (`^(?:[0-9a-fA-F]{2})*$`), not maximum length. Since `chainhash.HashB` handles arbitrary input, this is functionally safe, but a very large payload could be used for memory exhaustion. Consider adding a `max_length` constraint on the request field (e.g., 1MB).

4. **Missing `signer key not loaded` gRPC status mapping** — in `wallet/service.go`, when `SignerKey` is nil, the error is a plain `fmt.Errorf("signer key not loaded")`. The gRPC handler in `signer_handler.go` returns this raw error. It should probably be wrapped in `status.Errorf(codes.FailedPrecondition, ...)` for consistency with other handler methods.

### ✅ What looks good

- **Signature validation is thorough** — the infrastructure client (`signerClient`) checks both hex decode success and exact 64-byte length on the response.
- **Test coverage** — fixture-driven tests cover normal signing, empty message, 32-byte message, large message, different keys, and missing signer key. Verification round-trips through `schnorr.ParseSignature` + `Verify`.
- **Proto definition is clean** — input/output patterns are properly constrained in the OpenAPI spec.
- **gRPC handler hex-decode error** properly returns `codes.InvalidArgument`.

### 🟡 Nits

1. Missing newline at end of `signer_handler.go` and `service.proto` — will cause diff noise on next edit.
2. The OpenAPI spec title is empty in `options.openapi.json` — appears to be from the new `options.proto` file in #960, not this PR. Worth checking if that's intentional.

### Cross-repo impact
SDK parity: The `SignMessage` RPC is new. All SDKs (ts-sdk, go-sdk, rust-sdk, swift-sdk) will need to add the corresponding client method. Consider tracking this with an umbrella issue.

**Verdict:** Solid implementation. The two security nits (input length limit, gRPC status wrapping) are non-blocking but worth addressing.

**Review by ghost — COMMENTED** [review]
## 🔍 Arkana Review — Post-Update (1 new commit)

New commit: `d5c29636` — "check decoded signature is 64 bytes in SignMessage client"

### ✅ Good — Client-Side Signature Length Validation

The new check in `internal/infrastructure/signer/client.go`:

```go
if len(sig) != 64 {
    return nil, fmt.Errorf("invalid signature length: expected 64 bytes, got %d", len(sig))
}
```

This is good defensive coding. Schnorr signatures are always 64 bytes, and validating this on the client side catches malformed responses from the signer service early, before they propagate further into the signing pipeline.

This complements the OpenAPI spec validation (`^[0-9a-fA-F]{128}$` with min/max length 128) which enforces it at the HTTP layer. The Go client check catches issues in the gRPC path too.

### Previous Review Notes Still Apply

The overall SignMessage RPC looks clean:
- Message is hex-encoded, validated with `^(?:[0-9a-fA-F]{2})*$` regex
- Signing uses `chainhash.HashB(message)` (SHA-256 hash) then `schnorr.Sign` — standard Schnorr signing
- Tests cover valid signing, edge cases (empty/large messages), key selection, and missing-signer error
- No key material is leaked in responses

**No blocking issues.** The PR is in good shape after this fix.

**Review by louisinger — APPROVED** [review]

**Review by ghost — COMMENTED** [review]
**Iterative review** (merge commit `611db19c`)

Second merge of `master` into the branch. Changes are exclusively dependency bumps (Go 1.26.1→1.26.2, otel v1.40→v1.43, sys v0.41→v0.42) and a new error-level log flush in `telemetry/logrus_hook.go`. Zero diff in any SignMessage-related files.

No new concerns. Previous review stands — PR is clean and ready for final sign-off.

**Review by ghost — COMMENTED** [review]
## 🔍 Arkana Review — Post-Update (merge commit `a815e74b`)

Third merge of `master` into the branch. Conflict resolution in the three flagged files:

- **`service.pb.go`** — generated code, re-generated cleanly with the new `DeprecatedSigner` message types from master alongside the `SignMessage` types from this PR. Message indices and dep-idx tables are correct.
- **`service.go`** — `SignMessage` method properly placed after `LoadSignerKey` and before `Close`. No interaction with the new master code (deprecated-signer key lookup, effective-value coin selection, withdraw unlock-on-failure). Clean integration.
- **`signer_handler.go`** — `SignMessage` handler appended after `SignTransactionTapscript`. `GetPubkey` now returns deprecated signers (from master). No overlap with SignMessage logic.

No conflict markers. No drift in signing logic. SignMessage implementation unchanged from previously reviewed commits.

No new concerns. Previous review stands.

**Comment by altafan** [comment]
Not needed, we may reopen if realy required
